pkg/operator/configobservation/proxy/observe_proxy.go

```diff
@@ -0,0 +1,73 @@
+package proxy
+
+import (
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/klog"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/configobservation"
+	"github.com/openshift/library-go/pkg/operator/configobserver"
+	"github.com/openshift/library-go/pkg/operator/events"
+)
+
+func ObserveProxyConfig(genericListers configobserver.Listers, recorder events.Recorder, existingConfig map[string]interface{}) (map[string]interface{}, []error) {
+	const proxyPath = "proxy"
+
+	listers := genericListers.(configobservation.Listers)
+	errs := []error{}
+	prevObservedProxyConfig := map[string]interface{}{}
+
+	// grab the current Proxy config to later check whether it was updated
+	currentProxyMap, _, err := unstructured.NestedStringMap(existingConfig, proxyPath)
+	if err != nil {
+		return prevObservedProxyConfig, append(errs, err)
+	}
+
+	if len(currentProxyMap) > 0 {
+		unstructured.SetNestedStringMap(prevObservedProxyConfig, currentProxyMap, proxyPath)
+	}
+
+	observedConfig := map[string]interface{}{}
+	proxyConfig, err := listers.ProxyLister.Get("cluster")
+	if errors.IsNotFound(err) {
+		klog.Warningf("proxy.config.openshift.io/cluster: not found")
+		return observedConfig, errs
+	}
+	if err != nil {
+		return existingConfig, errs
+	}
+
+	newProxyMap := proxyToMap(proxyConfig)
+	if len(newProxyMap) > 0 {
+		if err := unstructured.SetNestedStringMap(observedConfig, newProxyMap, proxyPath); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if !reflect.DeepEqual(currentProxyMap, newProxyMap) {
+		recorder.Eventf("ObserveProxyConfig", "proxy changed to %q", newProxyMap)
+	}
+
+	return observedConfig, errs
+}
+
+func proxyToMap(proxy *configv1.Proxy) map[string]string {
+	proxyMap := map[string]string{}
+
+	if noProxy := proxy.Spec.NoProxy; len(noProxy) > 0 {
```

Comment:
do we need to validate these fields? "i made a typo in proxy and now my cluster is hosed"

Reply:
@danehans probably higher level question ^

Reply:
They are validated by the proxy controller inside of the network operator.